src/Microsoft.FeatureManagement/FeatureFilters/TimeWindowFilter.cs

```diff
@@ -56,7 +68,74 @@ public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext context)
+            //
+            // Hit the first occurrence of the time window
+            if ((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value))
+            {
+                return Task.FromResult(true);
+            }
+
+            if (settings.Recurrence != null)
+            {
+                //
+                // The reference of the object will be used for hash key.
+                // If there is no pre-bounded settings attached to the context, there will be no cached filter settings and each call will have a unique settings object.
+                // In this case, the cache for recurrence settings won't work.
+                if (context.Settings != null)
+                {
+                    DateTimeOffset cachedTime = _recurrenceCache.GetOrAdd(
+                        settings,
+                        (_) =>
+                        {
+                            if (RecurrenceEvaluator.TryFindPrevAndNextOccurrences(now, settings, out DateTimeOffset prevOccurrence, out DateTimeOffset _))
+                            {
+                                return prevOccurrence;
+                            }
+
+                            //
+                            // There is no previous occurrence within the reccurrence range.
+                            return DateTimeOffset.MaxValue;
+                        });
+
+                    if (now < cachedTime)
+                    {
+                        return Task.FromResult(false);
+                    }
+
+                    if (now < cachedTime + (settings.End.Value - settings.Start.Value))
+                    {
+                        return Task.FromResult(true);
+                    }
+
+                    if (RecurrenceEvaluator.TryFindPrevAndNextOccurrences(now, settings, out DateTimeOffset prevOccurrence, out DateTimeOffset nextOccurrrence))
+                    {
+                        bool isWithinPreviousTimeWindow =
+                            now <= prevOccurrence + (settings.End.Value - settings.Start.Value);
+
+                        _recurrenceCache.AddOrUpdate(
+                            settings,
+                            (_) => throw new KeyNotFoundException(),
+                            (_, _) => isWithinPreviousTimeWindow ? 
+                                prevOccurrence : 
+                                nextOccurrrence);
+
+                        return Task.FromResult(isWithinPreviousTimeWindow);
+                    }
+
+                    //
+                    // There is no previous occurrence within the reccurrence range.
+                    _recurrenceCache.AddOrUpdate(
+                        settings,
+                        (_) => throw new KeyNotFoundException(),
+                        (_, _) => DateTimeOffset.MaxValue);
+
+                    return Task.FromResult(false);
+                }
+
```

Comment:
I'd rather see the if reverted.

```
if (context.Settings == null)
{
    return Task.FromResult(RecurrenceEvaluator.MatchRecurrence(now, settings));
}

//
// Here evaluate with cached settings

//
// No longer need final `return Task.FromResult(false);` since it's handled as last line of whole method.

```